File: Carts/Animation/pop/CAAnimationStack.m
//
//  CAAnimationStack.m
//  GroceryList
//
//  Created by Hakon Hanesand on 3/28/15.

#import <pop/POP.h>
#import <ReactiveCocoa/ReactiveCocoa.h>

#import "CAAnimationStack.h"
#import "CAStoredAnimation.h"

#import "POPSpringAnimation+CAAdditions.h"
#import "POPAnimation+CAAnimation.h"

@implementation CAAnimationStack

- (void)pushAnimation:(POPSpringAnimation *)animation withTargetObject:(id)target andDescription:(NSString *)description {
    [target pop_addAnimation:animation forKey:description];
    CAStoredAnimation *reversedAnimation = [CAStoredAnimation animationWithSpring:[animation reverse] description:description targetObject:target];
    [self.stack addObject:reversedAnimation];
}

- (RACSignal *)popAnimation {
    CAStoredAnimation *topAnimation = [self.stack lastObject];
    [self.stack removeLastObject];
    [topAnimation startAnimation];
    return [topAnimation.animation completionSignal];
}

- (RACSignal *)popAllAnimations {
    return [[[self.stack.rac_sequence.signal doNext:^(CAStoredAnimation *animation) {
        [animation startAnimation];
    }] flattenMap:^RACStream *(CAStoredAnimation *animation) {
        return [animation.animation completionSignal];
    }] doCompleted:^{
        [self.stack removeAllObjects];
    }];
}

- (RACSignal *)popAnimationWithTargetObject:(id)target {
    return [[[self.stack.rac_sequence.signal filter:^BOOL(CAStoredAnimation *animation) {
        return [animation.targetObject isEqual:target];
    }] doNext:^(CAStoredAnimation *animation) {
        [animation startAnimation];
        [self.stack removeObject:animation];
    }] flattenMap:^RACStream *(CAStoredAnimation *animation) {
        return [animation.animation completionSignal];
    }];
}

- (NSMutableArray *)stack {
    if (!_stack) {
        _stack = [NSMutableArray new];
    }
    
    return _stack;
}

@end

File: Carts/Animation/pop/CAAnimationStack.swift
//
//  CAAnimationStack.swift
//  Carts
//
//  Created by Hakon Hanesand on 4/26/15.
//  Copyright (c) 2015 Hakon Hanesand. All rights reserved.
//

import UIKit

class CAAnimationStack: NSObject {
   
    var stack: Array<CAStoredAnimation>
    
    override required init() {
        stack = []
        super.init()
    }
    
    func pushAnimation(animation: POPSpringAnimation, targetObject: AnyObject, description: String) {
        targetObject.pop_addAnimation(animation, forKey: description)
        let reversedAnimation = CAStoredAnimation(spring: animation.reverse(), description: description, targetObject: targetObject)
        self.stack.append(reversedAnimation)
    }
    
    func popAnimation() -> RACSignal {
        let top = self.stack.removeLast()
        top.startAnimation()
        return top.animation.completionSignal()
    }
    
    func popAllAnimation() -> RACSignal {
        let array = NSMutableArray(array: self.stack)

        return array.rac_sequence.signal().doNext({ (next: AnyObject!) -> () in
            let ani = next as! CAStoredAnimation
            ani.startAnimation()
        }).flattenMap({ (next: AnyObject!) -> RACStream! in
            let ani = next as! CAStoredAnimation
            return ani.animation.completionSignal()
        }).doCompleted {
            self.stack.removeAll(keepCapacity: true)
        }
    }
    
    func popAnimationWithTargetObject(targetObject: AnyObject) -> RACSignal {
        let array = NSMutableArray(array: self.stack)
        
        return array.rac_sequence.signal().filter({ (next: AnyObject!) -> Bool in
            let ani = next as! CAStoredAnimation
            return ani.targetObject.isEqual(targetObject)
        }).doNext({ (next: AnyObject!) -> Void in
            let ani = next as! CAStoredAnimation
            ani.startAnimation()
            self.stack.removeAtIndex(find(self.stack, ani)!)
        }).flattenMap({ (next: AnyObject!) -> RACStream! in
            let ani = next as! CAStoredAnimation
            return ani.animation.completionSignal()
        })
    }
}
